### PR TITLE
Implemented push() and pop()

### PR DESCRIPTION
**Pull request by huttarl**
This pull requests implements push() and pop() methods. It also:
- updates the code to work with Processing 3.2.3
- uses float instead of int for location, rotation and related values
-- because of the above, split overloaded moveToward* methods into moveTowardRatio() and moveTowardDistance()

Not fully backward-compatible with all earlier sketches, because of change from int to float and renaming of moveToward() methods.

Tested with a new sketch that implements L-systems like https://en.wikipedia.org/wiki/L-system#Example_2:_Pythagoras_tree that require push and pop.

### CODE REVIEW

File: src/terrapin/HistoryTerrapin.java [hunk 1]
```diff
@@ -42,7 +42,8 @@
 	 */
 	public HistoryTerrapin(PApplet applet) {
 		super(applet);
-		applet.registerDraw(this);
+		// older versions of Processing: applet.registerDraw(this);
```
Comment (georgebashi) [inline src/terrapin/HistoryTerrapin.java]: You can drop this comment and just update for newer versions of Processing, I think.

File: src/terrapin/HistoryTerrapin.java [hunk 2]
```diff
@@ -53,7 +54,8 @@ public HistoryTerrapin(PApplet applet) {
 	 */
 	public HistoryTerrapin(Terrapin parent) {
 		super(parent);
-		applet.registerDraw(this);
+		// older versions of Processing: applet.registerDraw(this);
```
Comment (georgebashi) [inline src/terrapin/HistoryTerrapin.java]: same here, just drop the old version.

**Review by georgebashi — COMMENTED** [review]
There're a few changes here, and I have different thoughts about each. I'm very rusty with processing and this codebase, so excuse me if what I'm saying doesn't make sense.

Firstly, the conversion to float seems fine to me - something to be careful of (and the reason we went with int initially) is that due to floating point error people will easily hit cases where they will rotate some small amount in successive iterations (drawing a spirograph-like pattern), but the pattern won't end up joining up at the end. Forcing everything to ints neatly avoids this.

We intended this as a teaching library and hammered out the API based on a number of iterations in classrooms, and we found that gotchas like floating-point error were frustrating and confusing to beginners, and keeping newbies out of that space helped them to learn their way around, and we could gently introduce that later.

All that said, I'm not using this library at all anymore and given the effort you've put into it in this PR I feel I'm happy to call you the steward of it, so if you feel you'd rather go the other way on the tradeoff then that's totally reasonable and I'm happy to merge it.

Next with the push/pop thing: I think @ollieglass actually implemented the same thing at one point, but after some back-and-forth on it, my memory is that we decided that it provided some functionality that was pretty trivial to implement with a copy-constructor, which is why we added it. I think we then put that into our workshop, and used it as a way to teach how you could create one terrapin as a "template" and then fill an array with `new Terrapin(templateTerrapin)` copies and do fun things from there. We used it as a jumping off point to teach folks how to handle state, rather than taking the concept off their hands and providing the functionality in this API. `HistoryTerrapin` was an experiment in doing something similar and using it to teach how types and inheritance work.

Plus, we also wanted to keep the Terrapin class as super light-weight as possible so you could create millions of them, which is a fun thing for newbies to do.

I'm not really against the functionality in any way, but maybe it could live in a sub-class next to `HistoryTerrapin`? Like `TimeTravelTerrapin` or something? :)

Finally, your `moveToward` changes seem good. I did one of those "which idiot wrote this? oh wait, me." on that overloaded-method-doing-different-things disaster I'd created. I agree we should change the method names, but I'm not in love with `moveTowardDistance` and `moveTowardRatio` - I think something a little more verbose like `moveTowardByPixels` and `moveTowardByPct` might be more self-explanatory. That said, people generally used the `moveTowardByPct` version to implement following behaviors between turtles, so we could change it completely and call it `chase(Terrapin t, float ratio)` and keep the pixels one as `moveToward`.

Anyway, thanks for your work on this! I love that you're getting some use out of this code. I'd be super interested to hear what you're doing with it!

**Comment by huttarl** [comment]
Hey, great to hear back from you quickly, and with detailed comments. I appreciate your openness to suggestions.

It seems that there are different priorities depending on whether you're wanting to use the terrapin/turtle as an educational subject itself for students to learn programming with (as Logo was designed for), vs. as a platform for other kinds of graphics (which is more my use case) ... e.g. learning about L-systems. It was interesting to me that Leah Beuchley's turtle library (http://leahbuechley.com/Turtle/) specifically mentions the addition of push/pop as a reason for her library to exist when there was already Terrapin. In fact, I initially used her library... but I found it unwieldy for other reasons, and switched to Terrapin.

But I see your point about the push/pop stack not really belonging *in* the Terrapin class itself. And having the State in a separate class from the Terrapin, like I did it, seems redundant, whereas using instances of the Terrapin class to store state would be more elegant. So I'm going to do like you said and implement the stack with a separate class that manages copies of the Terrapin.

Regarding float/int: Definitely float is not without its problems. We could use double in order to reduce the rounding errors. But I found integers too limiting. E.g. I couldn't have a 22.5-degree turn. And if I had a turtle going forward 100 at a 60-degree angle, the rounding of its new location coordinates was a lot worse than the rounding error of a float. Also, if you had students who were supposed to be thinking in radians, you'd be forced to go with floating point anyway.

Regarding the names of the moveToward* methods... I agree that putting "By" in there helps make them a lot clearer.

I avoided "pixels" in naming the parameters and methods because it doesn't seem like the choice of pixels as the unit should be explicit; a turtle could be in a space where the units are cm, for example. Also, depending on the rotation, or device-dependent scaling (e.g. on a mobile device), Processing's units might end up not corresponding to pixels at all. It happens that in the sketch I'm using Terrapin for, I'm using scale() to fit designs to the canvas, so the coordinate system isn't in unitary pixels anyway. I guess at some point you have to get concrete about your units, but it doesn't seem necessary to hard-code it into the API's method names.

What about
    moveTowardByDistance()
    moveTowardByPercent()
?

I had chosen "ratio" over "percent" because "percent" implies a range of 0 to 100, whereas this parameter goes from 0 to 1; and the javadoc comment calls it a "ratio". But we could call it "percent" and just say that parameter should range from 0 to 100. Or leave it at "ratio" and continue to expect a parameter from 0 to 1.

**Pull request closed without merging**